### DISCUSSION TITLE
Using json with table valued function in Postgres

### DISCUSSION THREAD

**Discussion by amirotin** [sqlalchemy/sqlalchemy#7887] [Usage Questions] [answered]
Hi! I've complex sql query, which i'm trying to translate into sqlalchemy code. Main problem for me is connection between aliased tables and TableValuedAlias function. Actually i've few tables with json fields, which are connected through json values.
Instead of a lot of words, I'd rather show the code itself
This is query that i have:

```
select sum(
	(skill.row_data ->> 'base')::int + split_part(skill.row_data ->> 'leveled', '|', js."level")::int
)
from heroes, jsonb_to_recordset(skills) as js("level" int, "skillId" text), data_tables skill, data_tables buff
where skill.table_name = 'hero_skill'
and buff.table_name = 'effect_buff'
AND buff.row_id = (skill.row_data ->> 'effect')
AND (buff.row_data ->> 'type') = 'size'
and skill.row_id = js."skillId"
AND heroes.hero_id = 1
```

As you can see i have **heroes** table, two **data_tables** (aliased for different **table_name** value) and **jsonb_to_recordset** function. 

My models:
```
class DataTable(Base):
    __tablename__ = "data_tables"

    id = Column(Integer, primary_key=True, autoincrement=False)
    table_name = Column(String(50))
    row_id = Column(String(50))
    row_data = Column(JSON)

    Index('row_index', 'table_name', 'row_id', unique=True)

class Hero(Base):
    __tablename__ = "heroes"

    hero_id = Column(Integer, primary_key=True, autoincrement=False)
....
    skills = Column(JSONB, nullable=True)
```

This what i tried to make in sqlalchemy:

alias for **jsonb_to_recordset**
```
skills = aliased(func.jsonb_to_recordset(Hero.skills).table_valued(
    column("level", Integer), column("skillId", Text)
).render_derived(name="sk", with_types=True))
```
aliases for **DataTable**
```
skill = aliased(DataTable)
buff = aliased(DataTable)
```

and query
```
          stmt = select(
                func.sum(skill.row_data['base'].astext.cast(Integer) +
                         func.split_part(skill.row_data['leveled'].astext, '|', skills.c.level).cast(Integer))
            ).select_from(Hero, buff, skills).where(
                skill.table_name == 'hero_skill',
                buff.table_name == 'effect_buff',
                buff.row_id == skill.row_data['effect'].astext,
                buff.row_data['type'].astext == 'size',
                skill.row_id == skills.c.skillId,
                Hero.hero_id == 1
            )
```

Totally request turns out to be correct, with the exception of one moment - tables order in from clause:
```
SELECT sum(
	CAST((data_tables_1.row_data ->> %(row_data_1)s) AS INTEGER) + 
	CAST(split_part((data_tables_1.row_data ->> %(row_data_2)s), %(split_part_1)s, sk_1.level) AS INTEGER)
) AS sum_1 
FROM data_tables AS data_tables_1,
jsonb_to_recordset(heroes.skills) AS sk_1(level INTEGER, "skillId" TEXT),
data_tables AS data_tables_2,
heroes
WHERE data_tables_1.table_name = %(table_name_1)s
AND data_tables_2.table_name = %(table_name_2)s
AND data_tables_2.row_id = (data_tables_1.row_data ->> %(row_data_3)s) 
AND (data_tables_2.row_data ->> %(row_data_4)s) = %(param_1)s
AND data_tables_1.row_id = sk_1."skillId"
AND heroes.hero_id = %(hero_id_1)s
```

Because **heroes** is after **jsonb_to_recordset** i have error `missing FROM-clause entry for table "heroes"`
Perhaps I wrote the whole request incorrectly and I need to do something differently?

**Comment by zzzeek** [comment, marked as answer]
there's not really any ordering guarantee for FROM clauses right now, slightly surprising to me but I can see where it's doing this, so for now you'd have to work around this somehow.  if PG is fine with a JOIN on `true` you can do it like this:

```py
        .select_from(Hero, buff, skills).join_from(Hero, skills, true())
```

the SQL is:

```
SELECT sum(CAST((data_tables_1.row_data ->> :row_data_1) AS INTEGER) + CAST(split_part((data_tables_1.row_data ->> :row_data_2), :split_part_1, sk_1.level) AS INTEGER)) AS sum_1 
FROM data_tables AS data_tables_1, data_tables AS data_tables_2, heroes JOIN jsonb_to_recordset(heroes.skills) AS sk_1(level INTEGER, "skillId" TEXT) ON true 
WHERE data_tables_1.table_name = :table_name_1 AND data_tables_2.table_name = :table_name_2 AND data_tables_2.row_id = (data_tables_1.row_data ->> :row_data_3) AND (data_tables_2.row_data ->> :row_data_4) = :param_1 AND data_tables_1.row_id = sk_1."skillId" AND heroes.hero_id = :hero_id_1
```

we are getting a lot of table-valued issues like this, and it's not that hard to change the implementation for select() to maintain the ordering given in select_from() in most cases which might be useful.  at least it works here when i try it.  it's never been asked for.

**Reply by amirotin** [reply]
Actually, I thought I had something wrong with the use of table-valued function and aliased. I found a lot of examples on how to use it with raw json, but never how to use it with table column in sqlalchemy.
Your suggestion worked fo me! Thanks

**Reply by amirotin** [reply]
I found a way, to use subquery for **jsonb_to_recordset** function, but i got SAWarning `SELECT statement has a cartesian product between FROM element(s) "sk" and FROM element "heroes".  Apply join condition(s) between each element to resolve.`

```
skills = select(Hero.hero_id, func.jsonb_to_recordset(Hero.skills).table_valued(
                column("level", Integer), column("skillId", Text)
            ).render_derived(name="sk", with_types=True)).where(Hero.hero_id == 1)
```
There shouldn't be any join statements as i see. Why i get this warning?

**Reply by zzzeek** [reply]
the cartesian product linter doesnt automatically detect that jsonb_to_recordset() isn't going to generate a full set of rows independently.  use the [joins_implicitly](https://docs.sqlalchemy.org/en/14/core/functions.html?highlight=function#sqlalchemy.sql.functions.FunctionElement.alias.params.joins_implicitly) parameter (1.4.33 and above)

**Reply by amirotin** [reply]
I've 1.4.34 and joins_implicitly doesn't change anything

```
skills = select(Hero.hero_id, func.jsonb_to_recordset(Hero.skills).table_valued(
    column("level", Integer), column("skillId", Text), joins_implicitly=True
).render_derived(name="sk", with_types=True)).where(Hero.hero_id == 101)
logger.debug(skills)
result = session.execute(skills)
```

```
2022-04-02 21:50:16.267 | DEBUG    | on_hero_data:73 - SELECT heroes.hero_id, sk.level, sk."skillId" 
FROM heroes, jsonb_to_recordset(heroes.skills) AS sk(level INTEGER, "skillId" TEXT) 
WHERE heroes.hero_id = :hero_id_1
hero.py:74: SAWarning: SELECT statement has a cartesian product between FROM element(s) "sk" and FROM element "heroes".  Apply join condition(s) between each element to resolve.
  result = session.execute(skills)
```

does it work for jsonb_to_recordset?

**Reply by amirotin** [reply]
I could get it working properly only after small 'fix':

added `new_alias.joins_implicitly = self.joins_implicitly`
https://github.com/sqlalchemy/sqlalchemy/blob/990715adb4529432250c9864b586e59aa60ed372/lib/sqlalchemy/sql/selectable.py#L1602

Looks like when render_derived is called new TableValuedAlias is created and joins_implicitly passed to table _valued is ignored.

**Reply by zzzeek** [reply]
unfortunate but not surprising, thanks for reporting
